array_api_tests/test_creation_functions.py

```diff
@@ -130,9 +152,31 @@ def test_full(shape, fill_value, dtype):
     else:
         assert all(equal(a, asarray(fill_value, **kwargs))), "full() array did not equal the fill value"
 
-# TODO: implement full_like (requires hypothesis arrays support)
-def test_full_like():
-    pass
+@given(
+    a=xps.arrays(
+        dtype=shared_dtypes,
+        shape=shapes,
+    ),
+    fill_value=promotable_dtypes(shared_dtypes).flatmap(xps.from_dtype),
```

Comment:
Actually, this is the case for the other _like functions as well. If dtype=None, the dtype is the same as `x`. Otherwise the dtype is `dtype`, and the dtype of `x` is irrelevant. So in all these tests, the dtype of `x` (or `a` as we're calling it here, but we should change it. I like your idea of making the test signatures exactly match the signatures in the spec) should be drawn from all dtypes and be completely independent of the `dtype` parameter. So I think actually `shared_optional_promotable_dtypes` is not needed. We can just use `arrays(dtype=dtypes)` and `dtype=one_of(none(), dtypes)`.

Reply:
Oh I see now! I think I've fixed this now.

Using the suggested `@given(...)` I got a falsifying example with `(x=Array([], dtype=float32), fill_value=False, dtype=None)`, so I now generate `fill_value` from `x.dtype` if `dtype == None`.

For reference this is how I interpreted the spec:
```python
>>> from numpy import array_api as xp
>>> x = xp.zeros(1, dtype=xp.uint8)
>>> xp.full_like(x, 24)       # specified behaviour
Array([24], dtype=xp.uint8)
>>> xp.full_like(x, 4.2)      # unspecified behaviour
Array([4], dtype=xp.uint8) 
>>> xp.full_like(x, 256)      # unspecified behaviour
Array([0], dtype=uint8)
```

Let me know if the updated `test_full_like` is good.

Reply:
Here is the spec for full_like https://data-apis.org/array-api/latest/API_specification/creation_functions.html#full-like-x-fill-value-dtype-none-device-none. 

For the `dtype` parameter, it says "output array data type. If dtype is None , the output array data type must be inferred from fill_value (see full(shape, fill_value, *, dtype=None, device=None) ). Default: None ." The [`full` spec](https://data-apis.org/array-api/latest/API_specification/creation_functions.html#function-full) says "output array data type. If dtype is None , the output array data type must be inferred from fill_value . If the fill value is an int , the output array data type must be the default integer data type. If the fill value is a float , the output array data type must be the default floating-point data type. If the fill value is a bool , the output array must have boolean data type. Default: None ."

In both cases, the dtype of `x` is never mentioned. Only the value of `dtype`, or an inferred dtype from `fill_value` if `dtype=None`. So I think in your examples, the results should be 

```
>>> full_like(x, 24) # dtype=None
array(24, dtype=int64) OR array(24, dtype=int32)
>>> full_like(x, 4.2) # dtype=None
array(4.2, dtype=float64) OR array(4.2, dtype=float32)
>>> full_like(x, 2**100) # dtype=None
UNSPECIFIED
```

The dtype of `x` doesn't matter here. It could be anything. 

See https://data-apis.org/array-api/latest/API_specification/data_types.html#default-data-types. If dtype=None, the data type is the default floating-point dtype for float fill_value and the default integer dtype for int fill_value, which is either float32 or float64 and either int32 or int64, depending on the library. The only unspecified case is if the fill_value doesn't fit in the dtype. There's not really a clean way to infer the default dtype in the strategy, so we should just assume that the default is 32-bits for the purposes of generating fill_values within range when dtype=None.

So you're right that my @given isn't completely correct, because it doesn't account for unspecified fill values in the dtype=None case.

Reply:
If you're still unsure about this, let's discuss it on Slack. I want to make sure the spec is clear about what behavior it requires and that we are matching that here.

Reply:
Thanks for your explanation! I think I understand now, and have updated the test case accordingly.

So if I'm correct in saying, this is currently a bug in NumPy:
```python
>>> from numpy import array_api as xp
>>> x = xp.zeros(5, dtype=xp.int8)
>>> x
Array([0, 0, 0, 0, 0], dtype=int8)
>>> x_like = xp.full_like(x, fill_value=False)
>>> x_like
Array([0, 0, 0, 0, 0], dtype=int8)
```
i.e. `x_like` should be `Array([False, False, False, False, False, dtype=bool)`

Reply:
Ah you're right. This is why we have the test suite. I asked on slack to double check but it looks like this needs to be changed in NumPy. I'll make sure to include it in my next PR there.